hera_sim/visibilities/vis_cpu.py

```diff
@@ -356,50 +353,68 @@ def simulate(self, data_model):
-                    if self.polarized:
-                        vis_here = vis[p1, p2, :, ant1, ant2]
-                    else:
-                        vis_here = vis[:, ant1, ant2]
+            self._reorder_vis(
+                req_pols, data_model.uvdata, visfull[:, 0, i], vis, ant_list, polarized
+            )
 
-                    # get official "antenna numbers" corresponding to these indices
-                    antnum1, antnum2 = ant_list[ant1], ant_list[ant2]
+        # Reduce visfull array if in MPI mode
+        if self.mpi_comm is not None:
+            return self._reduce_mpi(visfull, myid)
 
-                    # get all blt indices corresponding to this antpair
-                    indx = data_model.uvdata.antpair2ind(antnum1, antnum2)
-                    if len(indx) == 0:
-                        # maybe we chose the wrong ordering according to the data. Then
-                        # we just conjugate.
-                        indx = data_model.uvdata.antpair2ind(antnum2, antnum1)
-                        vis_here = np.conj(vis_here)
+        return visfull
 
-                    visfull[indx, 0, i, p] = vis_here
+    def _reorder_vis(self, req_pols, uvdata, visfull, vis, ant_list, polarized):
+        indices = (
+            np.triu_indices(vis.shape[3])
+            if polarized
+            else np.triu_indices(vis.shape[1])
+        )
 
-        # Reduce visfull array if in MPI mode
-        if self.mpi_comm is not None:
-            from mpi4py.MPI import SUM
+        for p, (p1, p2) in enumerate(req_pols):
+            for ant1, ant2 in zip(*indices):  # go through indices in output
+                vis_here = (
+                    vis[p1, p2, :, ant1, ant2] if polarized else vis[:, ant1, ant2]
+                )
+                # get official "antenna numbers" corresponding to these indices
+                antnum1, antnum2 = ant_list[ant1], ant_list[ant2]
+
+                # get all blt indices corresponding to this antpair
+                indx = uvdata.antpair2ind(antnum1, antnum2)
+                if len(indx) == 0:
+                    # maybe we chose the wrong ordering according to the data. Then
+                    # we just conjugate.
+                    indx = uvdata.antpair2ind(antnum2, antnum1)
+                    vis_here = np.conj(vis_here)
+
+                visfull[indx, p] = vis_here
+
+    def _get_req_pols(self, uvdata, uvbeam, polarized: bool) -> List[Tuple[int, int]]:
+        if polarized:
+            x = uvbeam.feed_array.tolist().index("x")
+            y = uvbeam.feed_array.tolist().index("y")
+
+            avail_pols = {"xx": (x, x), "xy": (x, y), "yx": (y, x), "yy": (y, y)}
+            req_pols = []
+            for pol in uvdata.polarization_array:
+                # Get polarization strings in terms of n/e feeds
+                polstr = pyuvdata.utils.polnum2str(pol).lower()
+                req_pols.append(avail_pols[polstr])
+
+            return req_pols
```

Comment:
Based on the `UVBeam` docs, it looks like this isn't compatible with the full features of the class. This is fine if we're going to require users to have their beam objects only use `'x'` and `'y'` as feed labels, but maybe we want to be loud about it (e.g. issue a warning or raise an error) if that's what we're going to do. I've got an inkling of how to do this in a generalized way, so let me know if you want me to give it a try.

Reply:
Yeah, I was worried about that. We should get it right here, so if you have an idea of how to do it generally, go ahead!

Reply:
Here's what I'm thinking:
```
if not polarized:
    return [(0,0)]
else:
    feeds = list(uvbeam.feed_array)
    # in order to get all 4 visibility polarizations for a dual feed system
    vispols = set()
    for p1, p2 in itertools.combinations_with_replacement(feeds, 2):
        vispols.add(p1 + p2)
        vispols.add(p2 + p1)
    avail_pols = {
        vispol: (feeds.index(vispol[0]), feeds.index(vispol[1])) for vispol in vispols
    }
    # Get the mapping from uvdata pols to uvbeam pols
    req_pols = list(
        avail_pols[uvutils.polnum2str(polnum, uvbeam.x_orientation)]
        for polnum in uvdata.polarization_array
    )
    return req_pols    
```
If you suspect issues with the case of the feed labels and visibility polarizations, then we could throw in some `lower` methods, but I think this should work. The output should be a list with the same order as the data polarization array giving the indices to the corresponding feeds in the beam object.

Reply:
Hold on -- doesn't `polnum2str` only ever return something in `[xx, xy, yx, yy]`? What is the use-case you're thinking this covers?

Reply:
Only if the `x_orientation` parameter isn't specified. It's kind of hard to tell from the `pyuvdata` source code, but testing with version 2.1.5 shows that setting an `x_orientation` will output something in `['nn', 'en', 'ne', 'ee']`. The added bonus for this is that the data object doesn't care about what the beam object says its `x_orientation` is, since we're going from polarization numbers (which are `x_orientation` agnostic) to feed labels (which could change depending on whether an `x_orientation` is specified). Put another way, we don't actually need the `x_orientation` attribute to match up between the data and beam objects in order to sort the simulated data properly (though if we're trying to do our accounting properly, then maybe we set the data `x_orientation` to match the beam's after the fact). Does that make sense?

Reply:
I did some local testing, and this works whether the feed array uses the "xy" convention or the "en" convention. The original implementation implicitly requires that the feed array uses the xy convention. So to put it simply, this implementation covers both conventions (unless somehow the user is able to provide a beam object whose feeds are labeled with the "en" convention without specifying an `x_orientation`).

Reply:
Ah, I see! This is great -- can you cp/paste the code in?

Reply:
Done